### PR TITLE
Clarify link errors with CMake build

### PR DESCRIPTION
**Pull request by delcypher**
People seem to be running into these issues frequently so document how to work around it.

### CODE REVIEW

**Comment by naegling** [comment]
The proposed change probably would reduce build errors, but...
1) if users just want to run klee, they would be better off with the docker image (as stated in docs)
2) if users want to develop with klee, cmake is _much_ more flexible. And not that hard to configure ((build klee with no-rtti) or (build llvm with rtti)).
3) Users who do not fit into either cat 1 or 2 might be better off with autoconf, but I'm not entirely sure who they are... :)

**Comment by delcypher** [comment]
@naegling 
> if users want to develop with klee, cmake is much more flexible. And not that hard to configure ((build klee with no-rtti) or (build llvm with rtti)).

I'm not sure what your point is here. These changes are for the instructions for building **KLEE using CMake**.

**Comment by naegling** [comment]
Sorry, should have clarified. This comment in the proposed instructions:

  **NOTE:** If you build LLVM and Clang 3.4 from source **DO NOT USE CMAKE TO BUILD IT**. Use LLVM's Autoconf/Makefile build

I prefer to build my own LLVM so that I know exactly what I'm getting (rtti, ABI, etc). And 3.4 is old enough that it is not always available from your repo of choice. The above comment seems to imply that building llvm requires using autoconf.

wait,..., can you use autoconf for llvm and still use cmake for klee? I presumed you could not mix-and-match...

**Comment by delcypher** [comment]
> Sorry, should have clarified. This comment in the proposed instructions:
> 
>   **NOTE:** If you build LLVM and Clang 3.4 from source **DO NOT USE CMAKE TO BUILD IT**. Use LLVM's Autoconf/Makefile build
> 
> I prefer to build my own LLVM so that I know exactly what I'm getting (rtti, ABI, etc). And 3.4 is old enough that it is not always available from your repo of choice. The above comment seems to imply that building llvm requires using autoconf.
> 
> wait,..., can you use autoconf for llvm and still use cmake for klee? I presumed you could not mix-and-match...

Yes you can mix LLVM built with autoconf and KLEE build with CMake.

It was an explicit design for the CMake build system for KLEE that it be a **complete replacement for the old build system**. Consequently that meant building KLEE with CMake needed a way to build against LLVM built with either build system. The way it does this is to use the `llvm-config` binary to extract the information required.

This is where the problem with building LLVM/Clang with CMake comes in. The `llvm-config` binary when built with CMake  doesn't emit the `-fno-rtti` flag when it is supposed to. [This is a known bug and was fixed upstream](https://bugs.llvm.org/show_bug.cgi?id=14200) in r245064.

**Comment by delcypher** [comment]
@naegling sorry I accidentally edited your comment instead of replying to it **facepalm**. I've fixed it now.

It's clear this documentation needs some more work because it wasn't clear to you how you can mix and match build systems.

File: build-llvm34.md [hunk 1]
```diff
@@ -165,6 +169,42 @@ If you want to build KLEE with LLVM 2.9, [click here]({{site.baseurl}}/build-llv
...
+
+   This is caused by a mismatch between the ABI used to build LLVM and the ABI used to build KLEE. To fix this delete your KLEE build directory and rerun `cmake` like so
+
+   ```bash
+   $ CXXFLAGS="-D_GLIBCXX_USE_CXX11_ABI=0" cmake <CMAKE_OPTIONS> <KLEE_SRC_DIRECTORY>
+   ```
+
+   **NOTE:** If you see linker errors involving undefined references to `typeinfo` this is likely an [RTTI issue](https://github.com/klee/klee/issues/508).
+   Here's an example:
+   ```
```
Comment (andreamattavelli) [inline build-llvm34.md]: Missing empty line. Without it the rendering it incorrect (quite stupid, I know).

Comment (delcypher) [inline build-llvm34.md]: Fixed.

**Comment by delcypher** [comment]
@andreamattavelli Ping. Is this good to go?

**Comment by naegling** [comment]
Absolutely!

Cheers,
Rick



On 15 May 2017, at 05:57, Dan Liew <notifications@github.com<mailto:notifications@github.com>> wrote:


@andreamattavelli<https://github.com/andreamattavelli> Ping. Is this good to go?

**Comment by andreamattavelli** [comment]
Thanks @delcypher and sorry for the delay.

**Pull request merged**